### ISSUE TITLE
RyuJIT x86: GT_STORE_LCL_VAR decomposition is broken

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#6004]
I don't have a test case for this but the existing code is obviously wrong:
```
tree->gtNext = hiRhs;
hiRhs->gtPrev = tree;
hiRhs->gtNext = hiStore;
```

`tree` is actually `loStore` so this code basically takes the `hiRhs` node and moves it between `loStore` and `hiStore`. This fails whenever `hiRhs` is not a leaf. I suspect that this sometimes works due to the use of `fgNewStmtFromTree`, this function re-sequences the `hiRhs` tree and practically pulls it out from the old statement. But this doesn't happen in the case of embedded statements when `fgMakeEmbeddedStmt` is used instead.

I experimented with a workaround in #4986 - reverse the store ordering so `hiRhs` doesn't need to move. However, I'm far from being convinced that this is the right approach.

We could use temporaries again but it feels odd to use temporaries when decomposing `GT_STORE_LCL_VAR`.

**Comment by mikedn** [comment]
> We could use temporaries again but it feels odd to use temporaries when decomposing GT_STORE_LCL_VAR.

For operations with overflow checks we have no choice but to use a temporary or do the hi store before the lo store. With the current implementation the lo store is done before the hi half is computed and it's the hi half that usually throws in case of overflow.

**Comment by mikedn** [comment]
@RussKeldorph I'm working on this.

**Comment by BruceForstall** [comment]
@mikedn I think the LIR work in #6689 (hopefully to be merged in the next day or two) fixes this because it doesn't require maintaining tree order.

**Comment by mikedn** [comment]
@BruceForstall Yep. As noted in the related PR, we may still need to fix the way side effects are propagated to the 2 stores so I'll keep the PR open until we figure out what, if anything, is still needed from it.

**Comment by BruceForstall** [comment]
@mikedn. I believe this is now fixed, now that #6689 has been merged. If you agree, can you close this?

**Comment by mikedn** [comment]
@BruceForstall Yep, this should be closed. If there are problems with side effects I'll open a separate issue.
